server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/MaxRetryAllocationDecider.java

```diff
@@ -41,27 +41,36 @@
 
     public static final String NAME = "max_retry";
 
+    private static final Decision YES_NO_FAILURES = Decision.single(Decision.Type.YES, NAME, "shard has no previous failures");
+
     @Override
     public Decision canAllocate(ShardRouting shardRouting, RoutingAllocation allocation) {
         final UnassignedInfo unassignedInfo = shardRouting.unassignedInfo();
         final Decision decision;
-        if (unassignedInfo != null && unassignedInfo.getNumFailedAllocations() > 0) {
+        final boolean debug = allocation.debugDecision();
+        final int numFailedAllocations = unassignedInfo == null ? 0 : unassignedInfo.getNumFailedAllocations();
+        if (numFailedAllocations > 0) {
             final IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
             final int maxRetry = SETTING_ALLOCATION_MAX_RETRY.get(indexMetadata.getSettings());
-            if (unassignedInfo.getNumFailedAllocations() >= maxRetry) {
-                decision = allocation.decision(Decision.NO, NAME, "shard has exceeded the maximum number of retries [%d] on " +
-                    "failed allocation attempts - manually call [/_cluster/reroute?retry_failed=true] to retry, [%s]",
-                    maxRetry, unassignedInfo.toString());
-            } else {
-                decision = allocation.decision(Decision.YES, NAME, "shard has failed allocating [%d] times but [%d] retries are allowed",
-                    unassignedInfo.getNumFailedAllocations(), maxRetry);
-            }
+            final Decision res = numFailedAllocations >= maxRetry ? Decision.NO : Decision.YES;
+            decision = debug ? debugDecision(res, unassignedInfo, numFailedAllocations, maxRetry) : res;
         } else {
-            decision = allocation.decision(Decision.YES, NAME, "shard has no previous failures");
+            decision = debug ? YES_NO_FAILURES : Decision.YES;
```

Comment:
I am not sure we need to switch on `debug` for purely constant decisions?

Reply:
I looked into this and I think no we don't, we seem to only be using the full explanation in the explain allocation request -> will adjust accordingly